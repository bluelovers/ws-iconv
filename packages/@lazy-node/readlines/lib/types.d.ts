/// <reference types="node" />
/// <reference types="node" />
/**
 * Created by user on 2020/5/29.
 */
import { PathLike } from 'fs';
export interface IOptions {
    readChunk?: number;
    newLineCharacter?: INewLineCharacter;
}
export declare const enum EnumNewLineCharacter {
    LF = 10,
    CR = 13
}
export declare type IPathLike = PathLike | number;
export declare type INewLineCharacter = string | number | EnumNewLineCharacter | (string | number | EnumNewLineCharacter)[] | Buffer;
