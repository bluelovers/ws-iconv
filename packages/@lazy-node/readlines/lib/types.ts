/**
 * Created by user on 2020/5/29.
 */
import { PathLike } from 'fs';

export interface IOptions
{
	readChunk?: number;
	newLineCharacter?: INewLineCharacter;
}

export const enum EnumNewLineCharacter
{
	LF = 0x0a,
	CR = 0x0d,
}

export type IPathLike = PathLike | number
export type INewLineCharacter = string | number | EnumNewLineCharacter | (string | number | EnumNewLineCharacter)[] | Buffer

