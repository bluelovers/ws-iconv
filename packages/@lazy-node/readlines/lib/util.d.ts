/// <reference types="node" />
import { INewLineCharacter } from './types';
export declare function handleOptionNewLineCharacter(newLineCharacter?: INewLineCharacter): number[];
export declare function bufferEndWithByBuffer(buf: Buffer, value: Uint8Array, byteOffset?: number, encoding?: BufferEncoding): boolean;
export declare function bufferEndWith(buf: Buffer, value: string | number | Uint8Array, byteOffset?: number, encoding?: BufferEncoding): boolean;
export declare function bufferStripEndWithByBuffer(buf: Buffer, value: Uint8Array, byteOffset?: number, encoding?: BufferEncoding): Buffer;
export declare function bufferStripEndWith(buf: Buffer, value: string | number | Uint8Array, byteOffset?: number, encoding?: BufferEncoding): Buffer;
export declare function bufferIndexWithByBuffer(buf: Buffer, value: Uint8Array, byteOffset: number, encoding?: BufferEncoding): boolean;
export declare function bufferIndexWith(buf: Buffer, value: string | number | Uint8Array, byteOffset: number, encoding?: BufferEncoding): boolean;
export declare function splitBufferByBuffer(buffer: Buffer, value: Uint8Array): Buffer[];
