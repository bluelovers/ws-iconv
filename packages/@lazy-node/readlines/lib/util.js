"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitBufferByBuffer = exports.bufferIndexWith = exports.bufferIndexWithByBuffer = exports.bufferStripEndWith = exports.bufferStripEndWithByBuffer = exports.bufferEndWith = exports.bufferEndWithByBuffer = exports.handleOptionNewLineCharacter = void 0;
function handleOptionNewLineCharacter(newLineCharacter) {
    if (!newLineCharacter) {
        return [10 /* EnumNewLineCharacter.LF */];
    }
    else if (typeof newLineCharacter !== 'number') {
        let ls = [];
        for (let i = 0; i < newLineCharacter.length; i++) {
            let c = newLineCharacter[i];
            if (typeof c !== 'number') {
                ls.push(c.charCodeAt(0));
            }
            else {
                ls.push(c);
            }
        }
        return ls;
    }
    return [newLineCharacter];
}
exports.handleOptionNewLineCharacter = handleOptionNewLineCharacter;
function bufferEndWithByBuffer(buf, value, byteOffset, encoding) {
    let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
    return i !== -1 && (i + value.length) === buf.length;
}
exports.bufferEndWithByBuffer = bufferEndWithByBuffer;
function bufferEndWith(buf, value, byteOffset, encoding) {
    if (typeof value === 'number') {
        return buf[buf.length - 1] === value;
    }
    let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
    return i !== -1 && (i + value.length) === buf.length;
}
exports.bufferEndWith = bufferEndWith;
function bufferStripEndWithByBuffer(buf, value, byteOffset, encoding) {
    let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
    if (i !== -1 && (i + value.length) === buf.length) {
        return buf.slice(0, i);
    }
    return buf;
}
exports.bufferStripEndWithByBuffer = bufferStripEndWithByBuffer;
function bufferStripEndWith(buf, value, byteOffset, encoding) {
    if (typeof value === 'number') {
        if (buf[buf.length - 1] === value) {
            return buf.slice(0, buf.length - 1);
        }
    }
    else {
        let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
        if (i !== -1 && (i + value.length) === buf.length) {
            /*
            console.dir({
                i,
                vl: value.length,
                bl: buf.length,

                v: value.toString(),
                b: buf.toString(),
                b2: buf.slice(0, i).toString(),
            })
             */
            return buf.slice(0, i);
        }
    }
    return buf;
}
exports.bufferStripEndWith = bufferStripEndWith;
function bufferIndexWithByBuffer(buf, value, byteOffset, encoding) {
    let i = buf.indexOf(value, byteOffset, encoding);
    return i !== -1 && i === byteOffset;
}
exports.bufferIndexWithByBuffer = bufferIndexWithByBuffer;
function bufferIndexWith(buf, value, byteOffset, encoding) {
    if (typeof value === 'number') {
        return buf[byteOffset] === value;
    }
    let i = buf.indexOf(value, byteOffset, encoding);
    return i !== -1 && i === byteOffset;
}
exports.bufferIndexWith = bufferIndexWith;
function splitBufferByBuffer(buffer, value) {
    const lines = [];
    let bufferPosition = 0;
    let lastNewLineBufferPosition = 0;
    while (true) {
        let line;
        let bufferPositionValue = buffer[bufferPosition];
        if (bufferIndexWithByBuffer(buffer, value, bufferPosition++)) {
            line = buffer.slice(lastNewLineBufferPosition, bufferPosition += (value.length - 1));
            lines.push(line);
            lastNewLineBufferPosition = bufferPosition;
        }
        else if (bufferPositionValue === undefined) {
            break;
        }
    }
    let leftovers = buffer.slice(lastNewLineBufferPosition, bufferPosition);
    if (leftovers.length) {
        lines.push(leftovers);
    }
    return lines;
}
exports.splitBufferByBuffer = splitBufferByBuffer;
//# sourceMappingURL=util.js.map