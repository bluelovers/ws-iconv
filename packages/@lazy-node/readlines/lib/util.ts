import { INewLineCharacter, EnumNewLineCharacter } from './types';

export function handleOptionNewLineCharacter(newLineCharacter?: INewLineCharacter): number[]
{
	if (!newLineCharacter)
	{
		return [EnumNewLineCharacter.LF];
	}
	else if (typeof newLineCharacter !== 'number')
	{
		let ls = [] as number[]

		for (let i = 0; i < newLineCharacter.length; i++)
		{
			let c = newLineCharacter[i];

			if (typeof c !== 'number')
			{
				ls.push(c.charCodeAt(0))
			}
			else
			{
				ls.push(c)
			}
		}

		return ls
	}

	return [newLineCharacter]
}

export function bufferEndWithByBuffer(buf: Buffer,
	value: Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

	return i !== -1 && (i + value.length) === buf.length
}

export function bufferEndWith(buf: Buffer,
	value: string | number | Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	if (typeof value === 'number')
	{
		return buf[buf.length - 1] === value
	}

	let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

	return i !== -1 && (i + value.length) === buf.length
}

export function bufferStripEndWithByBuffer(buf: Buffer,
	value: Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

	if (i !== -1 && (i + value.length) === buf.length)
	{
		return buf.slice(0, i)
	}

	return buf
}

export function bufferStripEndWith(buf: Buffer,
	value: string | number | Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	if (typeof value === 'number')
	{
		if (buf[buf.length - 1] === value)
		{
			return buf.slice(0, buf.length - 1);
		}
	}
	else
	{
		let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

		if (i !== -1 && (i + value.length) === buf.length)
		{
			/*
			console.dir({
				i,
				vl: value.length,
				bl: buf.length,

				v: value.toString(),
				b: buf.toString(),
				b2: buf.slice(0, i).toString(),
			})
			 */

			return buf.slice(0, i)
		}
	}

	return buf
}

export function bufferIndexWithByBuffer(buf: Buffer,
	value: Uint8Array,
	byteOffset: number,
	encoding?: BufferEncoding,)
{
	let i = buf.indexOf(value, byteOffset, encoding)

	return i !== -1 && i === byteOffset
}

export function bufferIndexWith(buf: Buffer,
	value: string | number | Uint8Array,
	byteOffset: number,
	encoding?: BufferEncoding,)
{
	if (typeof value === 'number')
	{
		return buf[byteOffset] === value
	}

	let i = buf.indexOf(value, byteOffset, encoding)

	return i !== -1 && i === byteOffset
}

export function splitBufferByBuffer(buffer: Buffer, value: Uint8Array)
{
	const lines: Buffer[] = [];
	let bufferPosition = 0;

	let lastNewLineBufferPosition = 0;
	while (true)
	{
		let line: Buffer;
		let bufferPositionValue = buffer[bufferPosition];

		if (bufferIndexWithByBuffer(buffer, value, bufferPosition++))
		{
			line = buffer.slice(lastNewLineBufferPosition, bufferPosition += (value.length - 1));
			lines.push(line);
			lastNewLineBufferPosition = bufferPosition;
		}
		else if (bufferPositionValue === undefined)
		{
			break;
		}
	}

	let leftovers = buffer.slice(lastNewLineBufferPosition, bufferPosition);
	if (leftovers.length)
	{
		lines.push(leftovers);
	}

	return lines;
}
