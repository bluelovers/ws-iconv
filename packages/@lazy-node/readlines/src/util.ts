import { INewLineCharacter } from './types';
import { EnumNewLineCharacter } from './index';

export function handleOptionNewLineCharacter(newLineCharacter?: INewLineCharacter): number[]
{
	if (!newLineCharacter)
	{
		return [EnumNewLineCharacter.LF];
	}
	else if (typeof newLineCharacter !== 'number')
	{
		const ls = [] as number[];

		for (let i = 0; i < newLineCharacter.length; i++)
		{
			const c = newLineCharacter[i];

			if (typeof c !== 'number')
			{
				ls.push(c.charCodeAt(0))
			}
			else
			{
				ls.push(c)
			}
		}

		return ls
	}

	return [newLineCharacter]
}

