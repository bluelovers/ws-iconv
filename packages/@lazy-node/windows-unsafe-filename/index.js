"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.replaceWindowsUnsafeName = exports.hasWindowsUnsafeName = exports.newRegExpWindowsUnsafeName = void 0;
const windowsReservedRe = newRegExpWindowsUnsafeName();
const windowsReservedRe2 = newRegExpWindowsUnsafeName(true);
/**
 * @see https://stackoverflow.com/questions/38457621/what-happens-when-writing-to-aux-file-on-windows
 * @see https://github.com/parshap/node-sanitize-filename/blob/209c39b914c8eb48ee27bcbde64b2c7822fdf3de/test.js#L83
 */
function newRegExpWindowsUnsafeName(mode) {
    if (mode) {
        return /^(con|prn|aux|nul|com\d{1,2}|lpt\d{1,2})(\..*)?$/i;
    }
    return /^(con|prn|aux|nul|com[0-9]|lpt[0-9])(\..*)?$/i;
}
exports.newRegExpWindowsUnsafeName = newRegExpWindowsUnsafeName;
function hasWindowsUnsafeName(name, mode) {
    return name.match(mode ? windowsReservedRe2 : windowsReservedRe);
}
exports.hasWindowsUnsafeName = hasWindowsUnsafeName;
function replaceWindowsUnsafeName(name, replaceValue, mode) {
    return name.replace(mode ? windowsReservedRe2 : windowsReservedRe, replaceValue);
}
exports.replaceWindowsUnsafeName = replaceWindowsUnsafeName;
exports.default = hasWindowsUnsafeName;
//# sourceMappingURL=index.js.map