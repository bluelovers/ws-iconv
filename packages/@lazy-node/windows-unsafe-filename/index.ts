const windowsReservedRe = newRegExpWindowsUnsafeName();

const windowsReservedRe2 = newRegExpWindowsUnsafeName(true);

/**
 * @see https://stackoverflow.com/questions/38457621/what-happens-when-writing-to-aux-file-on-windows
 * @see https://github.com/parshap/node-sanitize-filename/blob/209c39b914c8eb48ee27bcbde64b2c7822fdf3de/test.js#L83
 */
export function newRegExpWindowsUnsafeName(mode?: boolean)
{
	if (mode)
	{
		return /^(con|prn|aux|nul|com\d{1,2}|lpt\d{1,2})(\..*)?$/i
	}

	return /^(con|prn|aux|nul|com[0-9]|lpt[0-9])(\..*)?$/i
}

export function hasWindowsUnsafeName(name: string, mode?: boolean)
{
	return name.match(mode ? windowsReservedRe2 : windowsReservedRe)
}

export function replaceWindowsUnsafeName(name: string,
	replaceValue: string | ((substring: string, ...args: (string | undefined)[]) => string),
	mode?: boolean,
)
{
	return name.replace(mode ? windowsReservedRe2 : windowsReservedRe, replaceValue as any)
}

export default hasWindowsUnsafeName
