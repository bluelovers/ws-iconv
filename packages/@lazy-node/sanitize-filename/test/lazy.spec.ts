import { inspect } from 'util';
import sanitizeFilename from '../index';
import getRandomString from 'get-random-string';
import { newRegExpWindowsUnsafeName } from '../../windows-unsafe-filename/index';
import { reZeroWidthTrim } from 'zero-width/lib/re';

describe(`describe`, () =>
{

	[
		"valid name.mp3",
		"résumé",
		"hello\u0000world",
		"hello\nworld",
		"h?w", "h/w", "h*w",
		"mr.", "mr..", "mr ", "mr  ",
		".", "..", "./", "../", "/..", "/../", "*.|.",
		"..",
		'a'.repeat(252) + '\uD800\uDC00',
		'\uD800\uDC00' + 'a'.repeat(252),
		'a'.repeat(253) + '\uD800\uDC00',
		'\uD800\uDC00' + 'a'.repeat(253),
	].forEach(input =>
	{
		test(inspect(input), () =>
		{
			let actual = sanitizeFilename(input);

			expect(actual).toMatchSnapshot();

			console.log(inspect(input) , '=>', inspect(actual))
		})
	});

})

describe(`random`, () =>
{
	const windowsReservedRe = newRegExpWindowsUnsafeName();

	Array(10)
		.fill(0)
		.map(() => {
			let s = getRandomString();
			let s2 = getRandomString({ regex: windowsReservedRe.source + '[<>:"/\\\\|?*\u0000-\u001F]' });

			return [
				s,
				s2,
			]
		}).flatMap(s => {
		return [
			s,
			`\uD800\uDC00${s}`,
			`${s}\uD800\uDC00`,
			`COM1.${s}`,
		]
	}).flat().forEach(input =>
	{
		test(inspect(input), () =>
		{
			let actual = sanitizeFilename(input);

			console.log(inspect(input) , '=>', inspect(actual))

			expect(actual.length > 0).toBeTruthy()
		})
	});

})
