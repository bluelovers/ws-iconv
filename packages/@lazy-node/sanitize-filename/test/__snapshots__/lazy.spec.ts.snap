// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe '*.|.' 1`] = `"."`;

exports[`describe '.' 1`] = `"!"`;

exports[`describe '..' 1`] = `"!"`;

exports[`describe '..' 2`] = `"!"`;

exports[`describe '../' 1`] = `"!"`;

exports[`describe './' 1`] = `"!"`;

exports[`describe '/..' 1`] = `"!"`;

exports[`describe '/../' 1`] = `".."`;

exports[`describe 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaêÄÄ' 1`] = `"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"`;

exports[`describe 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaêÄÄ' 1`] = `"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"`;

exports[`describe 'h*w' 1`] = `"h!w"`;

exports[`describe 'h/w' 1`] = `"h!w"`;

exports[`describe 'h?w' 1`] = `"h!w"`;

exports[`describe 'hello\\nworld' 1`] = `"helloworld"`;

exports[`describe 'hello\\x00world' 1`] = `"hello!world"`;

exports[`describe 'mr  ' 1`] = `"mr"`;

exports[`describe 'mr ' 1`] = `"mr"`;

exports[`describe 'mr.' 1`] = `"mr"`;

exports[`describe 'mr..' 1`] = `"mr"`;

exports[`describe 'r√©sum√©' 1`] = `"r√©sum√©"`;

exports[`describe 'valid name.mp3' 1`] = `"valid name.mp3"`;

exports[`describe 'êÄÄaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' 1`] = `"êÄÄaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"`;

exports[`describe 'êÄÄaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' 1`] = `"êÄÄaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"`;
